### ISSUE TITLE
Infoview DocumentPosition should be a context

### ISSUE THREAD

**Issue by EdAyers** [leanprover/vscode-lean4#197]
Currently in the infoview code, every component is passing the `pos : DocumentPosition` prop up through the component hierarchy. I think it would make a lot of sense to make this a React context instead.

**Comment by gebner** [comment]
I'm not sure if this is such a good idea.  We have a lot of different positions floating around.
1. The cursor position in the editor.
2. The position of the current info (which may be different from the cursor position if it's pinned).
3. The position of the data shown in the info (which may lag behind the info position when the cursor is moved).
4. The position used for RPC queries.
5. The position of the diagnostic message / term goal / widget.

It goes without saying that all of the can be different.  I vaguely recall that we've had some issues mixing up 2 and 3 but I can't find it right now.

**Comment by lovettchris** [comment]
I agree with Gabriel, I think "pinning" shows an example where a react component needs it's own local "pos" that is not in a shared context.  You see this in Infos.tsx where the pos being passed down to the <Info> component is conditionally coming from  the `curLoc` state or the `pinnedPoss` state. Infos is also listening to the `ec.events.changedCursorLocation` event so it can update when the current location is changed.  `AllMessages` is also an example of a React component in the info view that does *not* use the cursor location at all.

![image](https://user-images.githubusercontent.com/18707114/174671589-9b40ba61-22af-4f89-88ff-fb21e065e8a3.png)

I am curious about the naming of `pinnedPoss` and setPinnedPoss`.  I would have named these `pinnedPos` and setPinnedPos`.  I'm not sure what the second `s` at the and of those names represents exactly... it's not like it is plural positions or anything... so a bit confusing to me?  Or is it plural because of the `pinKey.current` refcount?

**Comment by EdAyers** [comment]
Yes there are different `pos`es flying around. I still think it's easier to manage these with contexts. Once you are inside the body of `InfoDisplay` there is only one pos. so the proposal would be to add pos as a context there.

**Comment by gebner** [comment]
> once you are inside the body of `InfoDisplay` there is only one pos

Inside of `InfoDisplay` you still have the positions (2)-(5) from my list.
